collect_sources.sh

```diff
@@ -63,23 +75,23 @@ EXCLUDE+='|raspberrypi-firmware|picontrol|revpi-firmware'
 # fetch Raspbian sources
 [ ! -d "$2" ] && mkdir -p "$2"
 cd "$2"
-dpkg-query --admindir $APTROOT/var/lib/dpkg -W		\
-	-f='${source:Package}=${source:Version}\n'	\
-	| egrep -v "^($EXCLUDE)=" | sort | uniq		\
-	| while read package ; do fetch_deb_src "$package" ; done
+dpkg-query --admindir $APTROOT/var/lib/dpkg -W \
+	-f='${source:Package}=${source:Version}\n' |
+	grep -E -v "^($EXCLUDE)=" | sort | uniq |
+	while read -r package; do fetch_deb_src "$package"; done
 
 # fetch RevolutionPi sources
 knl_version=$(dpkg-query --admindir $APTROOT/var/lib/dpkg -W \
 	-f='${source:Version}' raspberrypi-kernel || true)
 knl_tag="raspberrypi-kernel_$knl_version"
-knl_tag=$(sed 's/:/%25/g' <<< "$knl_tag")
+knl_tag=${knl_tag//\:/%25}
```

Comment:
It would be nice to add a source code comment here why we swap the : for %25. Makes reading and review easier.

Reply:
That's indeed confusing. I assume the idea is to escape the colon characters in the tag so that the result can be used verbatim in a URL. However, hexadecimal 25 is the % character (percent) in the ASCII table, not the : character (colon). That would be %3A.

Pull request otherwise LGTM.

Reply:
https://git-scm.com/docs/git-check-ref-format
> They cannot have ASCII control characters (i.e. bytes whose values are lower than \040, or \177 DEL), space, tilde ~, caret ^, or colon : anywhere.

https://dep-team.pages.debian.net/deps/dep14/
> Version mangling
>
>When a Git tag needs to refer to a specific version of a Debian package, the Debian version needs to be mangled to cope with Git's restrictions. This mangling is deterministic and reversible:
>
>    Each colon (:) is replaced with a percent (%)
>    Each tilde (~) is replaced with an underscore (_)
>    A hash (#) is inserted between each pair of adjacent dots (..)
>    A hash (#) is appended if the last character is a dot (.)
>    If the version ends in precisely .lock (dot l o c k, lowercase, at the end of the version), a hash (#) is inserted after the dot, giving .#lock.
>
>This can be expressed concisely in the following Perl5 statements:
>
> y/:~/%_/;
> s/\.(?=\.|$|lock$)/.#/g;
>
>The reverse transformation is:
>
>    Each percent (%) is replaced with a colon (:)
>    Each underscore (_) is replaced with a tilde (~)
>    Each hash (#) is deleted

Reply:
I think, I found an explanation: GIT tags cannot contain any of the following characters `\, ?, ~, ^, :, * , [, @`. Therefore the `:` is replaced with `:` (resp. the url-encoded ASCII representation), which we use for our official tags. I'll add a comment to above.

Reply:
@l1k  @naruxde I added a brief explanation in my last commit.